sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSourceV2.scala

```diff
@@ -0,0 +1,247 @@
+
+
+object TextSocketSourceProviderV2 {
+  val HOST = "host"
+  val PORT = "port"
+  val INCLUDE_TIMESTAMP = "includeTimestamp"
+  val NUM_PARTITIONS = "numPartitions"
+  val SCHEMA_REGULAR = StructType(StructField("value", StringType) :: Nil)
+  val SCHEMA_TIMESTAMP = StructType(StructField("value", StringType) ::
+    StructField("timestamp", TimestampType) :: Nil)
+  val DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
+}
+
+class TextSocketSourceProviderV2 extends DataSourceV2
+    with MicroBatchReadSupport with DataSourceRegister with Logging {
+  override def shortName(): String = "socketv2"
+
+  override def createMicroBatchReader(
+      schema: Optional[StructType],
+      checkpointLocation: String,
+      options: DataSourceV2Options): MicroBatchReader = {
+    logWarning("The socket source should not be used for production applications! " +
+      "It does not support recovery.")
+    if (!options.get(TextSocketSourceProviderV2.HOST).isPresent) {
+      throw new AnalysisException("Set a host to read from with option(\"host\", ...).")
+    }
+    if (!options.get(TextSocketSourceProviderV2.PORT).isPresent) {
+      throw new AnalysisException("Set a port to read from with option(\"port\", ...).")
+    }
+    if (schema.isPresent) {
+      throw new AnalysisException("The socket source does not support a user-specified schema.")
+    }
+
+    if (options.get(TextSocketSourceProviderV2.INCLUDE_TIMESTAMP).isPresent) {
+      Try(options.get(TextSocketSourceProviderV2.INCLUDE_TIMESTAMP).get().toBoolean) match {
+        case Success(bool) =>
+        case Failure(_) =>
+          throw new AnalysisException(
+            "includeTimestamp must be set to either \"true\" or \"false\"")
+      }
+    }
+
+    new TextSocketStreamMicroBatchReader(options)
+  }
+}
+
+case class TextSocketStreamOffset(offset: Long) extends Offset {
+  override def json(): String = offset.toString
+}
+
+class TextSocketStreamMicroBatchReader(options: DataSourceV2Options)
+  extends MicroBatchReader with Logging {
+
+  import TextSocketSourceProviderV2._
+
+  private var start: TextSocketStreamOffset = _
+  private var end: TextSocketStreamOffset = _
+
+  private val host = options.get(HOST).get()
+  private val port = options.get(PORT).get().toInt
+  private val includeTimestamp = options.getBoolean(INCLUDE_TIMESTAMP, false)
+  private val numPartitions = options.getInt(NUM_PARTITIONS, 1)
+
+  @GuardedBy("this")
+  private var socket: Socket = _
+
+  @GuardedBy("this")
+  private var readThread: Thread = _
+
+  @GuardedBy("this")
+  private val batches = new ListBuffer[(String, Timestamp)]
+
+  private val currentOffset = new AtomicLong(-1L)
+
+  private var initialized = false
+
+  @GuardedBy("this")
+  private var lastOffsetCommitted: Long = -1L
+
+  override def setOffsetRange(start: Optional[Offset], end: Optional[Offset]): Unit = {
+    if (!initialized) {
```

Comment:
Is it possible to initialize in the constructor?

Reply:
This is what I want to bring out. Originally I initialized this in constructor like old socket source. But I found that `MicroBatchReader` will be created in two different places with two objects. So initializing in constructor will create two sock threads and connectors. This is different from V1 source. In V1 source, we only created source once, but with V2 `MicroBatchReader` we will create two objects in two different places (one for schema), which means such side-affect actions in constructor will have two copies. Ideally we should only create this `MicroBatchReader` once.

Reply:
I don't think this will solve that problem, since each reader will just have its own initialize bit.

In general, I think it's fine if we do a bit of extra work. V1 sources do have to support being created multiple times (in e.g. restart scenarios), and the lifecycles of the two V2 readers being created here don't overlap. (We should be closing the tempReader created in DataStreamReader, though.)